File: LeetCode 77 39 40-Combination and Combination sum/combination_sum_distinct.go
package main

import "fmt"

func combinationSum(candidates []int, target int) [][]int {
	result := [][]int{}
	backtrack(candidates, target, 0, []int{}, &result)

	return result
}

func backtrack(candidates []int, target int, pos int, currentCombination []int, result *[][]int) {
	if target == 0 {
		newCombination := append([]int{}, currentCombination...)
		fmt.Println("new combination:", newCombination)
		*result = append(*result, newCombination)
		return
	}

	if pos == len(candidates) || target < 0 {
		return
	}

	for i := pos; i < len(candidates); i++ {
		currentNum := candidates[i]
		currentCombination = append(currentCombination, currentNum)
		backtrack(candidates, target-currentNum, i, currentCombination, result)
		currentCombination = currentCombination[:len(currentCombination)-1]
	}
}

func main() {
	candidates, target := []int{2, 3, 6, 7}, 7
	fmt.Println("combinations:", combinationSum(candidates, target))
}

/*
outout:

*/

File: LeetCode 77 39 40-Combination and Combination sum/combination_sum_distinct.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        
        def backtrack(pos, target, currentCombination):
            if target == 0:
                print("new combination", currentCombination)
                result.append(currentCombination[::])
                return
            
            if pos == len(candidates) or target < 0:
                return
            
            for i in range(pos, len(candidates)):
                currentElement = candidates[i]
                currentCombination.append(currentElement)
                backtrack(i, target-currentElement, currentCombination)
                currentCombination.pop()
            
        backtrack(0, target, [])
        
        return result

if __name__ == "__main__":
    solution = Solution()
    candidates, target = [2, 3, 6, 7], 7
    print("combinations", solution.combinationSum(candidates, target))


"""
output:
('new combination', [2, 2, 3])
('new combination', [7])
('combinations', [[2, 2, 3], [7]])
"""